spacy/scorer.py

```diff
@@ -295,20 +295,23 @@ def score_spans(
             # Find all predidate labels, for all and per type
             gold_spans = set()
             pred_spans = set()
-            # Special case for ents:
-            # If we have missing values in the gold, we can't easily tell
-            # whether our NER predictions are true.
-            # It seems bad but it's what we've always done.
-            if attr == "ents" and not all(token.ent_iob != 0 for token in gold_doc):
-                continue
             for span in getter(gold_doc, attr):
                 gold_span = (span.label_, span.start, span.end - 1)
                 gold_spans.add(gold_span)
                 gold_per_type[span.label_].add((span.label_, span.start, span.end - 1))
             pred_per_type = {label: set() for label in labels}
-            for span in example.get_aligned_spans_x2y(getter(pred_doc, attr)):
-                pred_spans.add((span.label_, span.start, span.end - 1))
-                pred_per_type[span.label_].add((span.label_, span.start, span.end - 1))
+            align_x2y = example.alignment.x2y
+            for pred_span in getter(pred_doc, attr):
+                indices = align_x2y[pred_span.start : pred_span.end].dataXd.ravel()
+                if len(indices):
+                    g_span = gold_doc[indices[0] : indices[-1]]
+                    # Check we aren't missing annotation on this span. If so,
+                    # our prediction is neither right nor wrong, we just
+                    # ignore it.
+                    if all(token.ent_iob != 0 for token in g_span):
```

Comment:
This is ents-specific scoring in a generic method, so it's breaking the `sents` scoring. I'm not sure whether you just want to create a separate `score_ents` method or work it in here. If neither `ents` or `sents` work with this as-is it's kind of bad, so maybe there's a way to redesign it so it works more easily. I do really want to be able to point to a generic scoring method for this kind of scoring rather than have too many special cases that make it hard to understand what the scorer is actually counting.

Reply:
Ah I forgot that this is a generic `score_spans`. Yeah I think we have different instincts in these scoring functions. I think making the code too generic here is very limiting. I suppose I'll just make my own entity scorer.